### PR TITLE
v1 of dark mode

### PR DESCRIPTION
**Pull request by akshay1502**

### CODE REVIEW

**Comment by akshay1502** [comment]
Capture1
Capture2

File: app/components/local-navbar.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import { action } from '@ember/object';
+import setCookie from '../helpers/setCookie';
+
+export default class ExampleComponent extends Component {
+  @action
+  changeTheme(event) {
+    const sunURL = 'assets/sun.png';
+    const moonURL = 'assets/moon.png';
+
+    if (event.target.src === 'https://dev.realdevsquad.com/assets/moon.png') {
```
Comment (harshith-venkatesh) [inline app/components/local-navbar.js:11]: Could we create a constant for the link and point it here please

File: app/styles/app.css [hunk 2]
```diff
@@ -6,22 +6,48 @@
...
+  --colorText: #E1E1EC;
+  --colorCardBg: #312F3B;
+  --taskButton: #3e3c3c;
+  --browseButton: #35393b;
+  --browseHover: #2b2f31;
+  --selectStatus: #181a1b;
+}
+
+.lightTheme {
+  --colorBackground: white;
```
Comment (harshith-venkatesh) [inline app/styles/app.css:23]: NIT: Shall we keep all in hexCode only?

**Review by harshith-venkatesh — CHANGES_REQUESTED** [review]

File: app/styles/app.css (hunk 2 above)
Comment (akshay1502) [reply app/styles/app.css]: done.

File: app/components/local-navbar.js (hunk 1 above)
Comment (akshay1502) [reply app/components/local-navbar.js:8]: I went for adding URL directly in src, because here, the URL will not change, but the image might get change. And even if we change the image, we will replace it with existing, so the URL will remain the same.

File: app/components/local-navbar.hbs [hunk 3]
```diff
@@ -11,4 +11,5 @@
   <span class="local-navbar__link">
     <LinkTo @route="notifications">Notifications</LinkTo>
   </span>
+  <img src='assets/moon.png' role="button" alt='light' id="themeLogo" class="themeLogo" {{on "click" this.changeTheme}}/>
```
Comment (swarajpure) [inline app/components/local-navbar.hbs:14]: Can we have more meaning text in "alt"?
Something like "Toggle the theme" or something?

File: app/components/local-navbar.js (hunk 1 above)
Comment (swarajpure) [reply app/components/local-navbar.js:8]: 1. Umm, you've set up dev.realdevsquad.com on your laptop/computer, but will the same link work on your phone?
2. Even according to your logic, if we tomorrow change the subdomain from `my.real......com` to `profile.real.....com`, will this URL be valid in that case?

File: app/helpers/setCookie.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { helper } from '@ember/component/helper';
+
+function setCookie(mode) {
+  const domain = '.realdevsquad.com';
+  const expire = new Date(Date.now() + 24 * 31 * 60 * 60 * 1000);
```
Comment (swarajpure) [inline app/helpers/setCookie.js:5]: Maybe call this `cookieExpiryDate`? 😂

File: app/helpers/setCookie.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  const cookie = `theme=${mode}; domain=${domain}; expires=${expire} path=/`;
+  document.cookie = cookie;
+
+  if (mode == 'light') {
```
Comment (swarajpure) [inline app/helpers/setCookie.js:9]: Why not triple equal?

File: app/styles/app.css [hunk 6]
```diff
@@ -6,22 +6,47 @@
 @import "local-navbar.css";
 @import "profile.css";
 
+.darkTheme {
+  --colorBackground: #1C1B22;
+  --colorLink: #0080FF;
+  --colorHighlight: #FF0090;
+  --colorText: #E1E1EC;
```
Comment (swarajpure) [inline app/styles/app.css:13]: This is cool!

File: app/styles/app.css [hunk 7]
```diff
@@ -6,22 +6,47 @@
...
+  --colorCardBg: #312F3B;
+  --taskButton: #3e3c3c;
+  --browseButton: #35393b;
```
Comment (swarajpure) [inline app/styles/app.css:16]: Do we have the variable colours documented somewhere?

File: .template-lintrc.js [hunk 8]
```diff
@@ -2,4 +2,9 @@
 
 module.exports = {
   extends: 'octane',
+
+  rules: {
+    'no-curly-component-invocation': { allow: ['setTheme'] },
+    'no-implicit-this': { allow: ['setTheme'] },
```
Comment (swarajpure) [inline .template-lintrc.js:8]: Could you please add reference to why was this added?

Comment (akshay1502) [reply .template-lintrc.js:8]: Screenshot (139)

File: app/components/local-navbar.js (hunk 1 above)
Comment (akshay1502) [reply app/components/local-navbar.js:8]: Ya. You are right. But when I logged event.target.src it showed `https://dev.realdevsquad.com/assets/moon.png` and if I just do `assets/moon.png` then it doesn't work.

File: app/styles/app.css (hunk 7 above)
Comment (akshay1502) [reply app/styles/app.css:16]: Yes. In the dark-mode channel, we have some primary colors documented. And for other colors, we used the Dark Reader extension to get colors.

File: app/components/local-navbar.hbs (hunk 3 above)
Comment (akshay1502) [reply app/components/local-navbar.hbs]: Good idea. Done.

File: app/components/local-navbar.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import setCookie from '../helpers/setCookie';
+import { SUN_URL, MOON_URL } from '../constants/url';
+
+export default class ExampleComponent extends Component {
+  @action
+  changeTheme(event) {
+    if (event.target.src.includes(MOON_URL)) {
```
Comment (swarajpure) [inline app/components/local-navbar.js:9]: Smooth!

Comment (swarajpure) [reply app/components/local-navbar.js:9]: Smartly handled! 🔥

File: app/helpers/setTheme.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import { helper } from '@ember/component/helper';
+import setCookie from './setCookie';
+
+function setTheme() {
+  const allCookies = document.cookie.split(';');
+  let obj = {};
+  for (let cookie of allCookies) {
+    const [key, value] = cookie.split('=');
+    obj = { ...obj, [key.trim()]: value };
```
Comment (swarajpure) [inline app/helpers/setTheme.js:9]: Not very clear with what's this doing
Maybe simplify the code? Or at least add comment as to what's it doing?

File: app/helpers/setTheme.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+  if (!Object.prototype.hasOwnProperty.call(obj, 'theme')) {
+    setCookie.compute('light');
+  } else {
+    if (obj.theme === 'dark') {
+      const img = document.getElementById('themeLogo');
+      img.src = 'assets/sun.png';
+      setCookie.compute('dark');
+    } else {
+      setCookie.compute('light');
```
Comment (swarajpure) [inline app/helpers/setTheme.js:20]: Could we also have `dark` and `light` coming from constants?

File: app/components/local-navbar.js (hunk 1 above)
Comment (akshay1502) [reply app/components/local-navbar.js]: Yes. Public -> assets -> moon.png

File: app/constants/theme.js [hunk 12]
```diff
@@ -0,0 +1,6 @@
+export const SUN_URL = 'assets/sun.png';
+export const MOON_URL = 'assets/moon.png';
+export const THEME = {
+  LIGHT: 'light',
+  DARK: 'dark',
+};
```
Comment (swarajpure) [inline app/constants/theme.js:6]: Perfect, thanks for this one!

File: app/helpers/setCookie.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import { helper } from '@ember/component/helper';
+
+function setCookie(mode) {
```
Comment (swarajpure) [inline app/helpers/setCookie.js:3]: What does `mode` tell us 🤔 
Maybe call it `theme` or `selectedTheme`?

File: app/helpers/setCookie.js [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+  const domain = '.realdevsquad.com';
+  const cookieExpiryDate = new Date(Date.now() + 24 * 31 * 60 * 60 * 1000);
+  const cookie = `theme=${mode}; domain=${domain}; expires=${cookieExpiryDate} path=/`;
+  document.cookie = cookie;
+
+  if (mode === 'light') {
```
Comment (swarajpure) [inline app/helpers/setCookie.js:9]: constants please :)

File: app/helpers/setCookie.js [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+    document.body.classList.remove('darkTheme');
+    document.body.classList.add('lightTheme');
+  } else {
+    document.body.classList.remove('lightTheme');
+    document.body.classList.add('darkTheme');
```
Comment (swarajpure) [inline app/helpers/setCookie.js:14]: @akshay1502 could we use `classList.toggle` here? https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle

File: app/helpers/setTheme.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+import setCookie from './setCookie';
+import { SUN_URL, THEME } from '../constants/theme';
+
+function setTheme() {
+  const getCookies = document.cookie.split(';');
+  let cookies = {};
+
+  //storing cookie as key value pair in Object
+  getCookies.forEach((cookie) => {
+    const [key, value] = cookie.split('=');
```
Comment (swarajpure) [inline app/helpers/setTheme.js:11]: Maybe add an example of `cookie` in the comment for future understanding?
something like:
`// example of cookie: rds-session=ajkfhdfkahsfdfds`

File: app/styles/app.css [hunk 17]
```diff
@@ -24,6 +37,17 @@ body {
...
+html, body {
+  display: flex;
+  flex-direction: column;
+  font-family: 'Roboto', sans-serif, serif;
+  font-size: 15px;
+  height: 100%;
+  background-color: var(--colorBackground);
+  color: var(--colorText);
+  min-height: 100vh;
+}
```
Comment (swarajpure) [inline app/styles/app.css:49]: Let's please keep html, body styles on the top, since they're at the top of the DOM tree hierarchy

File: app/helpers/setCookie.js (hunk 15 above)
Comment (akshay1502) [reply app/helpers/setCookie.js:14]: I tried using the toggle. But in js, we can't toggle two classes like one adding and one removing, both will be removed or added. Whereas Jquery supports adding and removing two classes using the toggle.

Comment (swarajpure) [reply app/helpers/setCookie.js]: Use two toggle statements?

Comment (akshay1502) [reply app/helpers/setCookie.js]: @swarajpure i tried but it doesn't work properly. So I stuck to my method.

File: app/helpers/initialTheme.js [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+import setTheme from './setTheme';
+import setCookie from './setCookie';
+import { SUN_URL, THEME } from '../constants/theme';
+
+function initialTheme() {
+  const getCookies = document.cookie.split(';');
+  let cookies = {};
+
+  //storing cookie as key value pair in Object
+  //eg. cookies = { rds-session: asdfjkl }
```
Comment (swarajpure) [inline app/helpers/initialTheme.js:11]: Good use of comments, thanks for this one!

File: app/helpers/setCookie.js [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import { helper } from '@ember/component/helper';
+
+function setCookie(theme) {
+  const domain = '.realdevsquad.com';
+  const cookieExpiryDate = new Date(Date.now() + 24 * 31 * 60 * 60 * 1000);
```
Comment (swarajpure) [inline app/helpers/setCookie.js:5]: Can we please seperate the calculation in a variable? Read more here: https://github.com/ryanmcdermott/clean-code-javascript#use-searchable-names

File: app/helpers/setTheme.js [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+import { THEME } from '../constants/theme';
+
+function setTheme(theme) {
+  if (theme === THEME.LIGHT) {
+    document.body.classList.add('lightTheme');
+    document.body.classList.remove('darkTheme');
+  } else {
+    document.body.classList.remove('lightTheme');
+    document.body.classList.add('darkTheme');
+  }
```
Comment (swarajpure) [inline app/helpers/setTheme.js:11]: Did we try `toggling` classes here?

File: public/api/tasks.json [hunk 21]
```diff
@@ -0,0 +1,11 @@
...
+  "tasks": [
+    {
+      "title": "Adding loading substate",
+      "purpose": "improving DX",
+      "startedOn": 1637148114,
+      "endsOn": 1639740111,
+      "status": "active"
+    }
+  ]
+}
```
Comment (swarajpure) [inline public/api/tasks.json:11]: Is this file required?

File: app/helpers/setCookie.js (hunk 19 above)
Comment (akshay1502) [reply app/helpers/setCookie.js:5]: I have bookmarked the repo. Thanks for sharing it.

File: public/api/tasks.json (hunk 21 above)
Comment (akshay1502) [reply public/api/tasks.json:11]: I was not assigned to any task through rds, due to which on tasks page I was not seeing any(blank) and issue was about showing loading message while API fetching is going on. So Ankush told me to mock the API. I guess we should keep it for such future cases.
What you think? @swarajpure

File: app/helpers/setTheme.js (hunk 20 above)
Comment (akshay1502) [reply app/helpers/setTheme.js:11]: https://github.com/Real-Dev-Squad/website-my/pull/160#discussion_r748691248
Yes. If a user is on dark mode and user reloads the page, then classes will get toggled due to which he will see light mode, to overcome this we have to write extra code for handling this. But by using `add` and `remove` we are handling it efficiently.

File: app/helpers/setCookie.js (hunk 19 above)
Comment (swarajpure) [reply app/helpers/setCookie.js:5]: Welcome!

Comment (swarajpure) [reply app/helpers/setCookie.js:5]: Though what I was asking for is
```suggestion
  const COOKIE_DEFAULT_EXPIRY = 31 * 24 * 60 * 60 * 1000; // 31 days
  const cookieExpiryDate = new Date(Date.now() + COOKIE_DEFAULT_EXPIRY);
```

Please note that the sequence of 24 and 31. Earlier I mistook it as 24 days 🤷🏻‍♂️

File: public/api/tasks.json (hunk 21 above)
Comment (swarajpure) [reply public/api/tasks.json:11]: Not required as a part of this PR, can be deleted

**Comment by akshay1502** [comment]
https://user-images.githubusercontent.com/55057874/147481577-29eb5a91-794d-4390-b124-264716e64317.mp4

**Review by rohan09-raj — COMMENTED** [review]
@akshay1502 @shubham-y  
- The color of navbar in dark mode is `#1d1283` but we have decided to use `#191970`
-  How about having this in dark mode for `Change Status` button ??
image

```CSS    
    font-size: 1.25rem;
    color: #fff;
    font-weight: 700;
    border: 3px solid #fff;
    padding: 20px;
    margin-bottom: 40px;
    border-radius: 10px;
    background: #e49504;
    cursor: pointer;
```
- Also can we change this blue to links-blue (#0080FF) as currently, it is dull from all the blues in dark mode??
- Bg of active text fields can `#e1e1ec` or any shade of gray
image
